include/swift/SIL/SILBuilder.h

```diff
@@ -40,6 +41,40 @@ class SILBuilder {
   /// recorded in this list.
   SmallVectorImpl<SILInstruction *> *InsertedInstrs = nullptr;
 
+  /// Maps opened archetypes to their definitions. If provided,
+  /// can be used by the builder. It is supposed to be used
+  /// only by SILGen or SIL deserializers.
+  SILOpenedArchetypesTracker *OpenedArchetypes = nullptr;
+
+  bool canProvideTypeDefs() const {
+    return OpenedArchetypes != nullptr;
+  }
+
+  // Build a set of typedef operands for an apply instruction. This is done by
+  // scanning the list of substitutions and collecting all opened archetypes.
+  void createApplyTypeDefs(SILInstruction *I, ArrayRef<Substitution> Subs) {
```

Comment:
Hmm.  Unfortunately, I think there's too much potential confusion with C++ to use the term "typedef" for this.